File: aspnet/src/RateLimiter.cs
using System;
using System.Runtime.Caching;
using System.Security.Cryptography;
using System.Text;

namespace FireWTWall
{
    /// <summary>
    /// Sliding-window rate limiter backed by MemoryCache.
    /// Thread-safe; shared across all IIS worker threads in the AppDomain.
    /// </summary>
    public sealed class RateLimiter
    {
        private readonly int          _windowSec;
        private readonly int          _maxRequests;
        private readonly int          _blockDurationSec;
        private static readonly MemoryCache _cache = new MemoryCache("waf_ratelimit");
        private static readonly object      _sync  = new object();

        public RateLimiter(RateLimitConfig config)
        {
            _windowSec        = config.WindowSec;
            _maxRequests      = config.MaxRequests;
            _blockDurationSec = config.BlockDurationSec;
        }

        public struct Result
        {
            public bool Allowed;
            public int  Remaining;
            public int  RetryAfter;
        }

        public Result Check(string ip)
        {
            string key   = "rl_" + Md5(ip);
            string bKey  = "bl_" + Md5(ip);
            long   now   = DateTimeOffset.UtcNow.ToUnixTimeSeconds();

            lock (_sync)
            {
                // Is this IP currently blocked?
                object blocked = _cache.Get(bKey);
                if (blocked != null)
                {
                    long until = (long)blocked;
                    int  retry = (int)Math.Max(0, until - now);
                    return new Result { Allowed = false, Remaining = 0, RetryAfter = retry };
                }

                // Fetch or create sliding window entry
                var entry = (long[])_cache.Get(key); // [start, count]
                if (entry == null || now - entry[0] >= _windowSec)
                {
                    entry = new long[] { now, 0 };
                }

                entry[1]++;

                // Store with double-window TTL
                var policy = new CacheItemPolicy
                {
                    AbsoluteExpiration = DateTimeOffset.UtcNow.AddSeconds(_windowSec * 2)
                };
                _cache.Set(key, entry, policy);

                if (entry[1] > _maxRequests)
                {
                    long until = now + _blockDurationSec;
                    _cache.Set(bKey, until, new CacheItemPolicy
                    {
                        AbsoluteExpiration = DateTimeOffset.UtcNow.AddSeconds(_blockDurationSec)
                    });
                    return new Result { Allowed = false, Remaining = 0, RetryAfter = _blockDurationSec };
                }

                int remaining = Math.Max(0, _maxRequests - (int)entry[1]);
                return new Result { Allowed = true, Remaining = remaining, RetryAfter = 0 };
            }
        }

        private static string Md5(string input)
        {
            using (var md5 = MD5.Create())
            {
                byte[] hash = md5.ComputeHash(Encoding.UTF8.GetBytes(input));
                return BitConverter.ToString(hash).Replace("-", "").ToLowerInvariant();
            }
        }
    }
}

File: python/firewtwall/rate_limiter.py
from __future__ import annotations
import hashlib
import threading
import time
from dataclasses import dataclass
from typing import Dict, Any

from .config import RateLimitConfig


@dataclass
class RateLimitResult:
    allowed: bool
    remaining: int
    retry_after: int


class RateLimiter:
    """
    Sliding-window rate limiter backed by an in-memory dict.
    Thread-safe. For multi-process deployments, swap _store for a Redis backend.
    """

    def __init__(self, config: RateLimitConfig):
        self._window_sec = config.window_sec
        self._max_requests = config.max_requests
        self._block_duration_sec = config.block_duration_sec
        self._store: Dict[str, Any] = {}
        self._lock = threading.Lock()

    def check(self, ip: str) -> RateLimitResult:
        key = "rl_" + self._md5(ip)
        bkey = "bl_" + self._md5(ip)
        now = time.time()

        with self._lock:
            # Blocked?
            blocked = self._store.get(bkey)
            if blocked and now < blocked:
                retry = int(blocked - now)
                return RateLimitResult(allowed=False, remaining=0, retry_after=retry)

            entry = self._store.get(key)
            if not entry or (now - entry["start"]) >= self._window_sec:
                entry = {"start": now, "count": 0}

            entry["count"] += 1
            self._store[key] = entry

            if entry["count"] > self._max_requests:
                self._store[bkey] = now + self._block_duration_sec
                return RateLimitResult(
                    allowed=False, remaining=0, retry_after=self._block_duration_sec
                )

            remaining = max(0, self._max_requests - entry["count"])
            return RateLimitResult(allowed=True, remaining=remaining, retry_after=0)

    @staticmethod
    def _md5(value: str) -> str:
        return hashlib.md5(value.encode()).hexdigest()
